proxied-token/proxied-token-spec.ini

```diff
@@ -0,0 +1,295 @@
+    #hashedLocation({COMPILER}, {_CREATOR}, .IntList) |-> CREATOR_ID
++requires:
+    andBool 0  <Int  NEW_MASTER_COPY_ID 
+
+[changeMasterCopy-failure]
+
+[changeMasterCopy-failure-1]
+; masterCopy with address 0 supplied
+msg_sender: CREATOR_ID
++requires:
+    andBool 0 ==Int NEW_MASTER_COPY_ID
+
+[changeMasterCopy-failure-2]
+; function called by address different from creator
++requires:
+    andBool #rangeAddress(NEW_MASTER_COPY_ID)
+    andBool 0  <Int  NEW_MASTER_COPY_ID
+
+[changeMasterCopy-failure-3]
+; non-zero call value
+callValue: CALL_VALUE
+msg_sender: CREATOR_ID
++requires:
+    andBool #rangeUInt(256, CALL_VALUE)
+    andBool 0  <Int  CALL_VALUE
+    andBool #rangeAddress(NEW_MASTER_COPY_ID)
+    andBool 0  <Int  NEW_MASTER_COPY_ID
+
+; ERC20 functions
+
+[totalSupply]
+callData: #abiCallData("totalSupply", .TypedArgs)
+
+[totalSupply-success]
+statusCode: _ => EVMC_SUCCESS
+output: _ => #asByteStackInWidth(TOTAL, 32)
++storage:
+    #hashedLocation({COMPILER}, {_TOTALTOKENS}, .IntList) |-> TOTAL
++requires:
+    andBool #rangeUInt(256, TOTAL)
+
+[totalSupply-failure]
+callValue: CALL_VALUE
++requires:
+    andBool #rangeUInt(256, CALL_VALUE)
+    andBool 0  <Int  CALL_VALUE
+
+[balanceOf]
+callData: #abiCallData("balanceOf", #address(OWNER))
++requires:
+    andBool #rangeAddress(OWNER)
+
+[balanceOf-success]
+statusCode: _ => EVMC_SUCCESS
+output: _ => #asByteStackInWidth(BAL, 32)
++storage:
+    #hashedLocation({COMPILER}, {_BALANCES}, OWNER) |-> BAL
++requires:
+    andBool #rangeUInt(256, BAL)
+
+[balanceOf-failure]
+callValue: CALL_VALUE
++requires:
+    andBool #rangeUInt(256, CALL_VALUE)
+    andBool 0  <Int  CALL_VALUE
+
+[allowance]
+callData: #abiCallData("allowance", #address(OWNER), #address(SPENDER))
++requires:
+    andBool #rangeAddress(OWNER)
+    andBool #rangeAddress(SPENDER)
+
+[allowance-success]
+statusCode: _ => EVMC_SUCCESS
+output: _ => #asByteStackInWidth(ALLOWANCE, 32)
++storage:
+    #hashedLocation({COMPILER}, {_ALLOWANCES}, OWNER SPENDER) |-> ALLOWANCE
++requires:
+    andBool #rangeUInt(256, ALLOWANCE)
+
+[allowance-failure]
+callValue: CALL_VALUE
++requires:
+    andBool #rangeUInt(256, CALL_VALUE)
+    andBool 0  <Int  CALL_VALUE
+
+[approve]
+callData: #abiCallData("approve", #address(SPENDER), #uint256(VALUE))
++requires:
+    andBool #rangeAddress(SPENDER)
+    andBool #rangeUInt(256, VALUE)
+
+[approve-success]
+statusCode: _ => EVMC_SUCCESS
+output: _ => #asByteStackInWidth(1, 32)
+log: _:List ( .List => ListItem(#abiEventLog(PROXY_ID, "Approval", #indexed(#address(RANDOM_ID)), #indexed(#address(SPENDER)), #uint256(VALUE))) )
++storage:
+    #hashedLocation({COMPILER}, {_ALLOWANCES}, RANDOM_ID SPENDER) |-> (_:Int => VALUE)
+
+[approve-failure]
+callValue: CALL_VALUE
++requires:
+    andBool #rangeUInt(256, CALL_VALUE)
+    andBool 0  <Int  CALL_VALUE
+
+[transfer]
+statusCode: _ => EVMC_SUCCESS
```

Comment:
Explain in a comment why return status is success even for most failure rules - because return value false is used instead to denote failure.

This is a difference from behavior recommended in the standard and should be noted in the readme. Please describe precisely the failure cases where functions return false instead of THROW.

Reply:
done